src/workerd/server/fallback-service.c++

```diff
@@ -0,0 +1,270 @@
+      break;
+    case ImportType::INTERNAL:
+      request.setType("internal");
+      break;
+  }
+  request.setSpecifier(specifier);
+  request.setReferrer(referrer);
+
+  if (rawSpecifier != nullptr) {
+    request.setRawSpecifier(rawSpecifier);
+  }
+
+  if (attributes.size() > 0) {
+    auto attrs = request.initAttributes(attributes.size());
+    int n = 0;
+    for (auto& attr: attributes) {
+      attrs[n].setName(attr.key);
+      attrs[n].setValue(attr.value);
+      n++;
+    }
+  }
+
+  auto payload = json.encode(request);
+
+  kj::Maybe<kj::String> jsonPayload;
+  bool redirect = false;
+
+  {
+    kj::Thread loaderThread(
+        [&address, &jsonPayload, payload = kj::mv(payload), &redirect, &specifier]() mutable {
+      try {
+        kj::AsyncIoContext io = kj::setupAsyncIo();
+        kj::HttpHeaderTable::Builder builder;
+        auto headerTable = builder.build();
+
+        auto addr = io.provider->getNetwork().parseAddress(address, 80).wait(io.waitScope);
+        auto client = kj::newHttpClient(io.provider->getTimer(), *headerTable, *addr, {});
+
+        kj::HttpHeaders headers(*headerTable);
+        headers.set(kj::HttpHeaderId::HOST, "localhost");
+
+        auto request = client->request(kj::HttpMethod::POST, "/", headers, payload.size());
+        {
+          // Write then drop the payload.
+          kj::mv(request.body)->write(payload.asPtr().asBytes()).wait(io.waitScope);
+        }
+
+        kj::HttpClient::Response resp = request.response.wait(io.waitScope);
+
+        if (resp.statusCode == 301) {
+          // The fallback service responded with a redirect.
+          KJ_IF_SOME(loc, resp.headers->get(kj::HttpHeaderId::LOCATION)) {
+            redirect = true;
+            jsonPayload = kj::str(loc);
+          } else {
+            KJ_LOG(ERROR, "Fallback service returned a redirect with no location", specifier);
+          }
+        } else if (resp.statusCode != 200) {
+          // Failed! Log the body of the response, if any, and fall through without
+          // setting jsonPayload to signal that the fallback service failed to return
+          // a module for this specifier.
+          auto payload = resp.body->readAllText().wait(io.waitScope);
+          KJ_LOG(ERROR, "Fallback service failed to fetch module", payload, specifier);
+        } else {
+          jsonPayload = resp.body->readAllText().wait(io.waitScope);
+        }
+      } catch (...) {
+        auto exception = kj::getCaughtExceptionAsKj();
+        KJ_LOG(ERROR, "Fallback service failed to fetch module", exception);
+      }
+    });
+  }
+
+  return handleReturnPayload(kj::mv(jsonPayload), redirect, specifier);
+}
+}  // namespace
+
+ModuleOrRedirect tryResolve(Version version,
+    ImportType type,
+    kj::StringPtr address,
+    kj::StringPtr specifier,
+    kj::StringPtr rawSpecifier,
+    kj::StringPtr referrer,
+    const kj::HashMap<kj::StringPtr, kj::StringPtr>& attributes) {
+  return version == Version::V1
```

Comment:
If you make this type template argument, you can make this branch constexpr

Reply:
It's also worth noting that this is used only in local-dev and then only for certain types of operations. Perf optimizations here are likely to be low-impact.